The python note book has two section.

1. topic modeling
2. multiclassification model

Topic modeling is adopted to lable the text data. The labled data is used to develop a multiclassification model.
we adopted several classification algoritham and checked there metrics.




In [2]:
import pandas as pd
import os

Loading the data.
The data is downloaded from https://www.kaggle.com/benhamner/nips-papers

In [3]:
papers = pd.read_csv("C:/Jyothi/Software training/Python/TOPIC modeling/papers.csv")


In [4]:
papers.columns

Index(['id', 'year', 'title', 'event_type', 'pdf_name', 'abstract',
       'paper_text'],
      dtype='object')

In [5]:
papers.shape

(7241, 7)

In [6]:
papers=papers.drop(["id","event_type",'pdf_name'],axis =1)

In [7]:
papers.head()

year                                              title          abstract  \
0  1987  Self-Organization of Associative Database and ...  Abstract Missing   
1  1987  A Mean Field Theory of Layer IV of Visual Cort...  Abstract Missing   
2  1988  Storing Covariance by the Associative Long-Ter...  Abstract Missing   
3  1994  Bayesian Query Construction for Neural Network...  Abstract Missing   
4  1994  Neural Network Ensembles, Cross Validation, an...  Abstract Missing   

                                          paper_text  
0  767\n\nSELF-ORGANIZATION OF ASSOCIATIVE DATABA...  
1  683\n\nA MEAN FIELD THEORY OF LAYER IV OF VISU...  
2  394\n\nSTORING COVARIANCE BY THE ASSOCIATIVE\n...  
3  Bayesian Query Construction for Neural\nNetwor...  
4  Neural Network Ensembles, Cross\nValidation, a...

In [8]:
papers.paper_text.head(20)

0     767\n\nSELF-ORGANIZATION OF ASSOCIATIVE DATABA...
1     683\n\nA MEAN FIELD THEORY OF LAYER IV OF VISU...
2     394\n\nSTORING COVARIANCE BY THE ASSOCIATIVE\n...
3     Bayesian Query Construction for Neural\nNetwor...
4     Neural Network Ensembles, Cross\nValidation, a...
5     U sing a neural net to instantiate a\ndeformab...
6     Plasticity-Mediated Competitive Learning\n\nTe...
7     ICEG Morphology Classification using an\nAnalo...
8     Real-Time Control of a Tokamak Plasma\nUsing N...
9     Real-Time Control of a Tokamak Plasma\nUsing N...
10    Learning To Play the Game of Chess\n\nSebastia...
11    Multidimensional Scaling and Data Clustering\n...
12    An experimental comparison\nof recurrent neura...
13    133\n\nTRAINING MULTILAYER PERCEPTRONS WITH TH...
14    Interference in Learning Internal\nModels of I...
15    Active Learning with Statistical Models\n\nDav...
16    A Rapid Graph-based Method for\nArbitrary Tran...
17    Ocular Dominance and Patterned Lateral\nCo

In [9]:
papers.paper_text.describe()

count          7241
unique         7237
top       









freq              2
Name: paper_text, dtype: object

In [10]:
import re

In [11]:
# Remove punctuation
papers['paper_text_processed'] = papers['paper_text'].map(lambda x: re.sub('[,\.!?]', '', x))
papers['paper_text_processed'] = papers['paper_text_processed'].map(lambda x: re.sub('[\n]', ' ', x))


In [12]:
papers['paper_text_processed'] = papers['paper_text_processed'].map(lambda x: x.lower())


In [13]:
papers['paper_text_processed'].head(20)

0     767  self-organization of associative database...
1     683  a mean field theory of layer iv of visual...
2     394  storing covariance by the associative lon...
3     bayesian query construction for neural network...
4     neural network ensembles cross validation and ...
5     u sing a neural net to instantiate a deformabl...
6     plasticity-mediated competitive learning  terr...
7     iceg morphology classification using an analog...
8     real-time control of a tokamak plasma using ne...
9     real-time control of a tokamak plasma using ne...
10    learning to play the game of chess  sebastian ...
11    multidimensional scaling and data clustering  ...
12    an experimental comparison of recurrent neural...
13    133  training multilayer perceptrons with the ...
14    interference in learning internal models of in...
15    active learning with statistical models  david...
16    a rapid graph-based method for arbitrary trans...
17    ocular dominance and patterned lateral con

In [14]:
from wordcloud import WordCloud

In [15]:
papers['paper_text_processed'].values.take(0)

'767  self-organization of associative database and its applications hisashi suzuki and suguru arimoto osaka university toyonaka osaka 560 japan abstract an efficient method of self-organizing associative databases is proposed together with applications to robot eyesight systems the proposed databases can associate any input with some output in the first half part of discussion an algorithm of self-organization is proposed from an aspect of hardware it produces a new style of neural network in the latter half part an applicability to handwritten letter recognition and that to an autonomous mobile robot system are demonstrated  introduction let a mapping f : x -+ y be given here x is a finite or infinite set and y is another finite or infinite set a learning machine observes any set of pairs (x y) sampled randomly from x x y (x x y means the cartesian product of x and y) and it computes some estimate j : x -+ y of f to make small the estimation error in some measure usually we say that:

In [16]:
long_string = ','.join(list(papers['paper_text_processed'].values))


In [ ]:
wordcloud = WordCloud(background_color="white", max_words=5000, contour_width=3, contour_color='steelblue')
# Generate a word cloud
wordcloud.generate(long_string)
# Visualize the word cloud
wordcloud.to_image()

converting the text data to the digital foem using bag of words. 

Topic Modeling is kind of unsuperbidsed machine learning algoritham to lable the text data.First step in this process is chossing the number of clusters/topics. If we have a domain knowledge we can do that. For example, in the example below its the articles published in a journal. The journals often will have the list of prior defined themes. Based on those theme we can define the number of topics.
They are two assumption under the topic modeling algorithams:
1. each topic is the mixture of all words each carrying a specific weightage (the shape of the output after the execution of the alogoritham is number of topics*total feature names) in this example this value is 5*113277
2. each document is the mixture of all topics each carrying a specific weight.We choose the topic with high weighage as lable of the document. (the shape of the output after the execution of the alogoritham is total number of documents*number of topics) in this example this value is 7241*5

There are various algorithams to conduct topic modeling include Latent Semantic Analysis (LSA/LSI), Probabilistic Latent Semantic Analysis (pLSA), and Latent Dirichlet Allocation (LDA). In this example we are adopting LDA.


In [18]:
from sklearn.feature_extraction.text import CountVectorizer
count_vectorizer = CountVectorizer(max_df=0.8, min_df=2, stop_words='english')

In [19]:
count_data_matrix = count_vectorizer.fit_transform((papers['paper_text_processed']))



In [20]:
count_data_matrix 

<7241x113227 sparse matrix of type '<class 'numpy.int64'>'
	with 6221117 stored elements in Compressed Sparse Row format>

In [21]:
from sklearn.decomposition import LatentDirichletAllocation as LDA

In [22]:
number_topics = 5


lda = LDA(n_components=number_topics, random_state=42)


In [23]:
count_fit =lda.fit(count_data_matrix)


In [24]:
len(count_vectorizer.get_feature_names())

113227

In [25]:
import random

for i in range(10):
    random_id = random.randint(0,len(count_vectorizer.get_feature_names()))
    print(count_vectorizer.get_feature_names()[random_id])

av2
laurence
amortizes
ljk
682687
darkand
film
dj1
536539
cll


In [26]:
first_topic = lda.components_[0]
first_topic

array([9.36286309e+02, 4.16322891e+02, 2.51404068e+01, ...,
       2.33407009e+01, 8.17368166e+00, 2.03271719e-01])

In [27]:
lda.components_

array([[9.36286309e+02, 4.16322891e+02, 2.51404068e+01, ...,
        2.33407009e+01, 8.17368166e+00, 2.03271719e-01],
       [2.21774024e+03, 5.69279575e+02, 8.73124352e+01, ...,
        2.23882719e+01, 1.19273526e+00, 4.19434131e+00],
       [1.70080356e+03, 1.32484571e+03, 1.58529871e+02, ...,
        2.01043832e-01, 2.00380432e-01, 2.00213449e-01],
       [7.43531025e+02, 3.74782997e+02, 8.48988947e+01, ...,
        2.16979164e-01, 2.21768893e-01, 2.01633277e-01],
       [4.05638867e+02, 4.39768824e+02, 1.61183921e+01, ...,
        8.53004110e-01, 2.11433752e-01, 2.00540242e-01]])

In [28]:
first_topic.shape

(113227,)

In [29]:
top_topic_words = first_topic.argsort()[-10:]
top_topic_words

array([ 69675,  80842,  40990,  69112,  35299,  80865, 101674,  29257,
        70342,  72646], dtype=int64)

In [30]:
for i in top_topic_words:
    print(count_vectorizer.get_feature_names()[i])

linear
optimal
convex
let
bound
optimization
theorem
algorithms
log
matrix


In [31]:
topic_values = lda.transform(count_data_matrix)
topic_values.shape

(7241, 5)

In [32]:
papers['Topic'] = topic_values.argmax(axis=1)

In [33]:
papers.head()

year                                              title          abstract  \
0  1987  Self-Organization of Associative Database and ...  Abstract Missing   
1  1987  A Mean Field Theory of Layer IV of Visual Cort...  Abstract Missing   
2  1988  Storing Covariance by the Associative Long-Ter...  Abstract Missing   
3  1994  Bayesian Query Construction for Neural Network...  Abstract Missing   
4  1994  Neural Network Ensembles, Cross Validation, an...  Abstract Missing   

                                          paper_text  \
0  767\n\nSELF-ORGANIZATION OF ASSOCIATIVE DATABA...   
1  683\n\nA MEAN FIELD THEORY OF LAYER IV OF VISU...   
2  394\n\nSTORING COVARIANCE BY THE ASSOCIATIVE\n...   
3  Bayesian Query Construction for Neural\nNetwor...   
4  Neural Network Ensembles, Cross\nValidation, a...   

                                paper_text_processed  Topic  
0  767  self-organization of associative database...      4  
1  683  a mean field theory of layer iv of visual...      3  
2  394  storing covariance by the associative lon...      3  
3  bayesian query construction for neural network...      1  
4  neural network ensembles cross validation and ...      1

In [34]:
from sklearn.feature_extraction.text import TfidfVectorizer


In [35]:
TFIDF = TfidfVectorizer()
text_TF = TFIDF.fit_transform(papers["paper_text"])

In [36]:
text_TF.get_shape

<bound method spmatrix.get_shape of <7241x233506 sparse matrix of type '<class 'numpy.float64'>'
	with 7453306 stored elements in Compressed Sparse Row format>>

In [37]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    text_TF, papers['Topic'], test_size=0.3, random_state=123)

In [38]:
X_train.get_shape

<bound method spmatrix.get_shape of <5068x233506 sparse matrix of type '<class 'numpy.float64'>'
	with 5230482 stored elements in Compressed Sparse Row format>>

Developing models using the lableed data

In [39]:
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import SGDClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.naive_bayes import GaussianNB
from xgboost import XGBClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.multiclass import OneVsRestClassifier



In [40]:
model_name = ['LogisticRegression','RandomForestClassifier','AdaBoostClassifier','K Nearest Neighbor','XGBoost']
ac_score_list =[]
p_score_list=[]
r_score_list=[]
f1_score_list =[]

In [41]:
from sklearn import metrics
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score

In [42]:
%%time

m1=OneVsRestClassifier(LogisticRegression()).fit(X_train, y_train)
y_pred = m1.predict(X_test)
ac_score_list.append(accuracy_score(y_test, y_pred))
p_score_list.append(precision_score(y_test, y_pred, average='macro'))
r_score_list.append(recall_score(y_test, y_pred, average='macro'))
f1_score_list.append(f1_score(y_test, y_pred, average='macro'))

C:\Users\kesav\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Wall time: 5.91 s


In [56]:
f1_score_list

[0.884068589620238,
 0.6118549634312835,
 0.7143641527144888,
 0.6680189191568877,
 0.8409698042766827]

In [44]:
%%time

m2=RandomForestClassifier(random_state=3).fit(X_train, y_train)
y_pred = m2.predict(X_test)
ac_score_list.append(accuracy_score(y_test, y_pred))
p_score_list.append(precision_score(y_test, y_pred, average='macro'))
r_score_list.append(recall_score(y_test, y_pred, average='macro'))
f1_score_list.append(f1_score(y_test, y_pred, average='macro'))

C:\Users\kesav\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


Wall time: 2.42 s


In [52]:
y_pred.take(10)

1

In [47]:
%%time

m3=AdaBoostClassifier(random_state=3).fit(X_train, y_train)
y_pred = m3.predict(X_test)
ac_score_list.append(accuracy_score(y_test, y_pred))
p_score_list.append(precision_score(y_test, y_pred, average='macro'))
r_score_list.append(recall_score(y_test, y_pred, average='macro'))
f1_score_list.append(f1_score(y_test, y_pred, average='macro'))

Wall time: 1min 27s


In [50]:
%%time

m4= KNeighborsClassifier().fit(X_train, y_train)
y_pred = m4.predict(X_test)
ac_score_list.append(accuracy_score(y_test, y_pred))
p_score_list.append(precision_score(y_test, y_pred, average='macro'))
r_score_list.append(recall_score(y_test, y_pred, average='macro'))
f1_score_list.append(f1_score(y_test, y_pred, average='macro'))

Wall time: 8.53 s


In [53]:
%%time
m5=XGBClassifier().fit(X_train, y_train)
y_pred = m5.predict(X_test)
ac_score_list.append(accuracy_score(y_test, y_pred))
p_score_list.append(precision_score(y_test, y_pred, average='macro'))
r_score_list.append(recall_score(y_test, y_pred, average='macro'))
f1_score_list.append(f1_score(y_test, y_pred, average='macro'))

Wall time: 10min 44s


In [54]:
model_comparison_df = pd.DataFrame([model_name, ac_score_list, p_score_list, r_score_list, f1_score_list]).T
model_comparison_df.columns = ['model_name', 'accuracy_score', 'precision_score', 'recall_score', 'f1_score']
model_comparison_df = model_comparison_df.sort_values(by='f1_score', ascending=False)


In [55]:
model_comparison_df.head()

model_name accuracy_score precision_score recall_score  f1_score
0  LogisticRegression       0.886332        0.888675     0.880722  0.884069
4            AdaBoost       0.840313        0.844064     0.838506   0.84097
2      Decsision Tree       0.710078        0.714677     0.714779  0.714364
3       Random Forest       0.673263        0.737904     0.650032  0.668019
1       SGDClassifier       0.617579        0.649605     0.602568  0.611855

In [87]:
! pip freeze > requirements.txt

In [88]:

f = open("requirements.txt", "r")
print(f.read())

absl-py==0.9.0
alabaster==0.7.12
anaconda-client==1.7.2
anaconda-navigator==1.9.12
anaconda-project==0.8.2
asn1crypto==0.24.0
astor==0.8.1
astroid==2.1.0
astropy==3.1
astunparse==1.6.3
atomicwrites==1.2.1
attrs==18.2.0
Babel==2.6.0
backcall==0.1.0
backports.os==0.1.1
backports.shutil-get-terminal-size==1.0.0
beautifulsoup4==4.6.3
bitarray==0.8.3
bkcharts==0.2
blaze==0.11.3
bleach==3.0.2
bokeh==1.0.2
boto==2.49.0
Bottleneck==1.2.1
certifi==2020.6.20
cffi==1.11.5
chardet==3.0.4
Click==7.0
cloudpickle==0.6.1
clyent==1.2.2
colorama==0.4.1
comtypes==1.1.7
conda==4.8.4
conda-build==3.17.6
conda-package-handling==1.7.0
conda-verify==3.1.1
contextlib2==0.5.5
cryptography==2.4.2
cycler==0.10.0
Cython==0.29.2
cytoolz==0.9.0.1
dask==1.0.0
datashape==0.5.4
decorator==4.3.0
defusedxml==0.5.0
distributed==1.25.1
docutils==0.14
entrypoints==0.2.3
et-xmlfile==1.0.1
fastcache==1.0.2
filelock==3.0.10
Flask==1.0.2
Flask-Cors==3.0.7
future==0.17.1
gast==0.3.3
gevent==1.3.7
glob2==0.6
google-pasta==0.2.0
g

In [63]:

os.getcwd()

'C:\\Jyothi\\Software training\\Python\\TOPIC modeling'

In [62]:
os.chdir("C:/Jyothi/Software training/Python/TOPIC modeling")

In [91]:
os.remove("environment")

PermissionError: [WinError 5] Access is denied: 'environment'

In [90]:
file = open('requirement.txt','w')

file.write(f.read())

file.close()